### PR TITLE
[Torchscript] Adds Sequence and Text feature postprocessing

### PR DESCRIPTION
**Pull request by geoffreyangus**
This PR implements Sequence and Text postprocessing. In order to do this, we additionally standardize the outputs of both Sequence and Text output features. Now, both features output the following:

- `prediction`: the maximum token for each timestep
- `probabilities`: the probability of the maximum token for each timestep
- `probability`: the log-probability of the predicted sequence

```python
{
    "predictions": [
        ["lQhz", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB"],
        ["lQhz", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB", "HxgtdB"],
    ],
    "probabilities": tensor(
        [
            [0.2930, 0.2792, 0.2555, 0.2349, 0.2203, 0.2100, 0.2030, 0.1983, 0.1954],
            [0.2899, 0.2953, 0.2627, 0.2376, 0.2212, 0.2103, 0.2031, 0.1984, 0.1955],
        ]
    ),
    "probability": tensor(
        [
            -13.2349,
            -13.1438,
        ]
    ),
}

```

### CODE REVIEW

File: ludwig/features/feature_utils.py [hunk 1]
```diff
@@ -48,21 +48,50 @@ def set_str_to_idx(set_string, feature_dict, tokenizer_name):
...
+    else:
+        raise ValueError(f"probabilities type must be in [list, tuple, np.ndarray]. Got {type(probabilities)}")
+    return max_probs
+
+
+def compute_sequence_probability(
+    sequence_probabilities: np.ndarray,
+    max_sequence_length: Optional[int] = None,
+    return_log_prob: bool = True,
+):
```
Comment (justinxzhao) [inline ludwig/features/feature_utils.py:78]: nit: missing return type, i.e. `float`?

File: ludwig/features/sequence_feature.py [hunk 2]
```diff
@@ -121,6 +123,42 @@ def forward(self, v: TorchscriptPreprocessingInput):
...
+        pred_predictions = preds[self.predictions_key]
+        predictions: List[List[str]] = []
+        for sequence in pred_predictions:
+            sequence_predictions: List[str] = []
+            for i in range(self.max_sequence_length):
+                unit_id = int(sequence[i].item())
+                if unit_id < len(self.idx2str):
+                    unit_prediction = self.idx2str[unit_id]
+                else:
+                    unit_prediction = self.unknown_symbol
```
Comment (justinxzhao) [inline ludwig/features/sequence_feature.py:147]: When/Does the model actually predict `unit_id`s larger than `idx2str`?

IIRC, `idx2str` should already contain all of the special symbols, which includes the unknown symbol.

**Review by justinxzhao — COMMENTED** [review]
Nice consolidation. Had a few minor comments.

File: ludwig/features/feature_utils.py [hunk 3]
```diff
@@ -48,21 +48,50 @@ def set_str_to_idx(set_string, feature_dict, tokenizer_name):
     return np.array(out, dtype=np.int32)
 
 
-def compute_sequence_probability(sequence_probabilities: Union[list, tuple, np.ndarray]) -> float:
-    """Computes sequence-level probability.
+def compute_token_probabilities(
+    probabilities: Union[list, tuple, np.ndarray],
+) -> float:
```
[the remark this answers is not in the document]
Comment (geoffreyangus) [reply ludwig/features/feature_utils.py:53]: Done.

File: ludwig/features/feature_utils.py (hunk 1 above)
Comment (geoffreyangus) [reply ludwig/features/feature_utils.py:78]: Done.

File: ludwig/features/feature_utils.py [hunk 4]
```diff
@@ -48,21 +48,50 @@ def set_str_to_idx(set_string, feature_dict, tokenizer_name):
...
+def compute_sequence_probability(
+    sequence_probabilities: np.ndarray,
+    max_sequence_length: Optional[int] = None,
+    return_log_prob: bool = True,
+):
+    """Computes the sequence level probability.
+
+    Args:
+        sequence_probabilities: An iterable of iterables or np.ndarray with shape (sequence_length,)
+        max_sequence_length: The maximum sequence length to use. If None, use the length of the longest sequence.
```
[the remark this answers is not in the document]
Comment (geoffreyangus) [reply ludwig/features/feature_utils.py:83]: Done.

**Review by justinxzhao — APPROVED** [review]

**Pull request merged**